src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -139,30 +141,65 @@ UpdateTaskQueryString(Query *query, Oid distributedTableId, RangeTblEntry *value
 		valuesRTE->values_lists = task->rowValuesLists;
 	}
 
-	/*
-	 * For INSERT queries, we only have one relation to update, so we can
-	 * use deparse_shard_query(). For UPDATE and DELETE queries, we may have
-	 * subqueries and joins, so we use relation shard list to update shard
-	 * names and call pg_get_query_def() directly.
-	 */
+	bool localExecution = TaskAccessesLocalNode(task);
+
 	if (query->commandType == CMD_INSERT)
 	{
-		deparse_shard_query(query, distributedTableId, task->anchorShardId, queryString);
+		/*
+		 * For INSERT queries, we only have one relation to update, so we can
+		 * use deparse_shard_query().
+		 */
+		if (localExecution)
```

Comment:
Even if taskAccessesLocalNode, users can disable local execution by `citus.enable_local_execution` (or Citus sets this within a TX block when necessary).

Do we produce accurate results when the GUC is off?

Reply:
Since you have made tests for this on your branch I will resolve this.